tty/pc-tty/ttypc_mouse.c

```diff
@@ -0,0 +1,322 @@
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+
+#include "board_config.h"
+#include "ttypc_mouse.h"
+
+#if PC_TTY_ENABLE_MOUSE
+
+#include <poll.h>
+#include <unistd.h>
+
+#include <sys/file.h>
+#include <sys/interrupt.h>
+#include <sys/io.h>
+#include <sys/threads.h>
+
+#include <libklog.h>
+#include <posix/utils.h>
+
+#include "event_queue.h"
+#include "ttypc_ps2.h"
+
+
+#define POOL_THREAD_PRIORITY 1
+
+
+static int _ttypc_mouse_setup(ttypc_t *ttypc)
+{
+	unsigned char status;
+	int err;
+
+	do {
+		/* Enable mouse. Should receive ACK (0xFA) afterwards */
+		err = ttypc_ps2_write_ctrl(ttypc, 0xA8);
+		if (err < 0) {
+			break;
+		}
+		err = ttypc_ps2_read(ttypc);
+		if (err != 0xFA) {
+			(void)fprintf(stderr, "pc-tty: mouse not found\n");
+			err = -1;
+			break;
+		}
+
+		/* Get compaq status byte */
+		err = ttypc_ps2_write_ctrl(ttypc, 0x20);
+		if (err < 0) {
+			break;
+		}
+
+		status = ttypc_ps2_read(ttypc);
+		status |= (1u << 1u);  /* Enable IRQ12 interrupt */
+		status &= ~(1u << 5u); /* Enable mouse clock (unset bit 5) */
+
+		/* Set compaq status byte */
+		err = ttypc_ps2_write_ctrl(ttypc, 0x60);
+		if (err < 0) {
+			break;
+		}
+		err = ttypc_ps2_write(ttypc, status);
+		if (err < 0) {
+			break;
+		}
+
+		/* Set defaults. Should receive ACK afterwards */
+		err = ttypc_ps2_write_ctrl(ttypc, 0xD4);
+		if (err < 0) {
+			break;
+		}
+		err = ttypc_ps2_write(ttypc, 0xF6);
+		if (err < 0) {
+			break;
+		}
+		err = ttypc_ps2_read(ttypc);
+		if (err != 0xFA) {
+			(void)fprintf(stderr, "pc-tty: mouse initialization failed - no ACK after setting defaults\n");
+			err = -1;
+			break;
+		}
+
+		/* Send mouse cmd - enable packet streaming. Should receive ACK afterwards */
+		err = ttypc_ps2_write_ctrl(ttypc, 0xD4);
+		if (err < 0) {
+			break;
+		}
+		err = ttypc_ps2_write(ttypc, 0xF4);
+		if (err < 0) {
+			break;
+		}
+		err = ttypc_ps2_read(ttypc);
+		if (err != 0xFA) {
+			(void)fprintf(stderr, "pc-tty: mouse initialization failed - no ACK after enable packet streaming cmd\n");
+			err = -1;
+			break;
+		}
+
+		/* Is the mouse alive? It is, if ACK received */
+		err = ttypc_ps2_write_ctrl(ttypc, 0xD4);
+		if (err < 0) {
+			break;
+		}
+		err = ttypc_ps2_write(ttypc, 0xEB);
+		if (err < 0) {
+			break;
+		}
+		err = ttypc_ps2_read(ttypc);
+		if (err != 0xFA) {
+			(void)fprintf(stderr, "pc-tty: mouse died\n");
+			err = -1;
+		}
+	} while (0);
+
+	if (err < 0) {
+		(void)fprintf(stderr, "pc-tty: failed to initialize mouse\n");
+		return -1;
+	}
+
+	return EOK;
+}
+
+
+static void _ttypc_kbd_mouse_poolthr(void *arg)
+{
+	ttypc_t *ttypc = (ttypc_t *)arg;
+	msg_rid_t rid;
+	msg_t msg;
+
+	for (;;) {
+		if (msgRecv(ttypc->mport, &msg, &rid) < 0) {
+			continue;
+		}
+
+		if (libklog_ctrlHandle(ttypc->mport, &msg, rid) == 0) {
+			/* msg has been handled by libklog */
+			continue;
+		}
+
+		switch (msg.type) {
+			case mtOpen:
+				msg.o.err = EOK;
+				break;
+
+			case mtRead:
+				if (msg.o.size < 3) {
```

Comment:
Is reading a multiple of 3 bytes acceptable? Also, please adjust the formatting of the comment.